mempool/reactor_test.go

```diff
@@ -333,6 +333,45 @@ func TestReactorTxSendersMultiNode(t *testing.T) {
 	require.Zero(t, len(firstReactor.txSenders))
 }
 
+// Test the experimental feature that limits the number of outgoing connections for gossiping
+// transactions.
+func TestMempoolReactorMaxActiveOutboundConnections(t *testing.T) {
+	config := cfg.TestConfig()
+	config.Mempool.ExperimentalMaxUsedOutboundPeers = 1
+	reactors, _ := makeAndConnectReactors(config, 4)
+	defer func() {
+		for _, r := range reactors {
+			if err := r.Stop(); err != nil {
+				assert.NoError(t, err)
+			}
+		}
+	}()
+	for _, r := range reactors {
+		for _, peer := range r.Switch.Peers().List() {
+			peer.Set(types.PeerStateKey, peerState{1})
+		}
+	}
+
+	// Add a bunch transactions to the first reactor.
+	txs := newUniqueTxs(100)
+	callCheckTx(t, reactors[0].mempool, txs)
+
+	// Wait for all txs to be in the mempool of the second reactor; the third and fourth reactor
+	// should not receive any tx.
+	checkTxsInMempool(t, txs, reactors[1], 0)
+	require.Zero(t, reactors[2].mempool.Size())
+	require.Zero(t, reactors[3].mempool.Size())
+
+	// In the first reactor, disconnect the second reactor; the third reactor should become active.
+	firstPeer := reactors[0].Switch.Peers().List()[0]
+	reactors[0].Switch.StopPeerGracefully(firstPeer)
+
+	// Now the third reactor should receive the transactions; the fourth reactor's mempool should
+	// still be empty.
+	checkTxsInMempool(t, txs, reactors[2], 0)
```

Comment:
How are you sure that is the second reactor (and not the third) that will get the semaphore's permit?

Reply:
Yes, this test is quite brittle, it depends on the current implementation of the p2p layer. We know it works like that because of the order in which peers are created and connected, but this can change and brake the test.